### ISSUE TITLE
Ability to inspect the primitive type used to store data

### ISSUE THREAD

**Issue by chadrik** [schematics/schematics#429]
As a developer using schematics to automatically convert a model to an open schema format (e.g. json-schema), it would be very useful to be able to inspect the python data type used to store the primitive data (i.e. one of `str`, `int`, `float`, `bool`, `dict`, or `list`).  I propose a class attribute named `PRIMITIVE_TYPE`.  This is not something that would be required for a custom model to implement in order operate correctly, but it would be great for the built-in models to set the standard for users to follow.  

It can also be used to create variation through inheritance, as in:

``` python
class NumberType(BaseType):
    PRIMITIVE_TYPE = None
    ...

class IntType(NumberType):
    PRIMITIVE_TYPE = int
    ...

class FloatType(NumberType):
    PRIMITIVE_TYPE = float
    ...
```

**Comment by bintoro** [comment]
+1. I've thought about this too.

`UnionType` probably shouldn't advertise this information at all. The main reason is that, in order to inspect the inner type of a compound type, the inspecting code would need to retrieve the underlying compound type instance anyway. So we might as well keep things simple and say:

* `PRIMITIVE_TYPE` is always a type (and not, say, a tuple of possibilities)
* `UnionType` fields should be examined by inspecting the underlying types directly.

**Comment by chadrik** [comment]
Great. I'll make the pull request. Do you prefer caps for this variable?  There is some room for debate over whether abstract class attributes are static or not.

**Comment by bintoro** [comment]
Hmm. This raises an important question: is the primitive type always static, or is it allowed to vary according to, say, a parameter on the field or the value being exported?

I was able to find one case where the type depends on the value: the new `TimestampType` returns an `int` if the value has a zero microsecond component; otherwise a `float` is returned. I think this can be classified as a bug — a `float` should always be returned. But is this kind of behavior always a bug?

Unless someone can provide a legitimate use case, I'd be inclined to rule that the primitive type cannot depend on context-specific circumstances such as the value itself or options specified at export time, as that would be incompatible with static schema descriptions.

As another example, consider a variant of `IPv4Type` that exports the value as an integer instead of a dot-decimal string (let's pretend it doesn't inherit from `StringType`). Would it be mandatory to implement this as a separate type instead of a parameter on `IPv4Type`? If not, then `PRIMITIVE_TYPE` may sometimes have to be a property and should not look like a constant.

Assigning exactly one primitive datatype to every Schematics type certainly feels like a good idea. For example, we have `UTCDateTimeType` instead of an option on `DateTimeType`. But I'm not yet sure whether mandating such arrangement is *always* appropriate.

And finally, it may be a reasonable expectation that the native type be inspectable as well, in which case the same considerations apply.

**Comment by chadrik** [comment]
> Unless someone can provide a legitimate use case, I'd be inclined to rule that the primitive type cannot depend on context-specific circumstances such as the value itself or options specified at export time, as that would be incompatible with static schema descriptions.

I agree. For the purposes of generating a json-schema doc, I need the primitive type to be accessible from the class, so it can't vary based on instance data.

> And finally, it may be a reasonable expectation that the native type be inspectable as well, in which case the same considerations apply.

From a user of type classes, it also seems the most sane solution.  The alternative is that users must be aware of the cases where the returned type will change and adapt their code to predict/detect these and cope with it. 

Another question is whether we should enforce that this attribute exists and is not `None` in the `TypeMeta` class, and assert that the result of `to_primitive` matches the declared type.  Or should we simply document this as a best practice?

**Comment by lkraider** [comment]
+1 from me. It's a useful annotation, and transforming into a property should not cause problems if the user requires that in their custom types.

**Comment by chadrik** [comment]
> transforming into a property should not cause problems if the user requires that in their custom types

@lkraider, keep in mind that it would need to be a "classproperty" in order to provide class-level access.  Something like this:

```python
class classproperty(object):
    def __init__(self, func):
        self.func = func
    def __get__(self, inst, klass):
        return self.func(klass)

class MyType(BaseType):
    @classproperty
    def PRIMITIVE_TYPE(cls):
        return str if cls.whatever else int
```

**Comment by bintoro** [comment]
Why can't it vary based on instance data? Schemas are composed of instances of types.

I can't imagine a classproperty being very useful because its output would be essentially fixed. Might as well make it a constant in that case.

**Comment by chadrik** [comment]
> Why can't it vary based on instance data? 

I see what you're saying.  I think will be fine for me, but let me look into it.  Anyway, if that's the case, then it should definitely be lowercase `primitive_type`.

> I can't imagine a classproperty being very useful because its output would be essentially fixed. Might as well make it a constant in that case.

It can be handy to compute something based on an inherited value that you want accessible at the class-level:

```python
class MyBaseClass(object):
    @classproperty
    def type_id(cls):
        return "{}.{}".format(cls.__module__, cls.__name__)
```

**Comment by bintoro** [comment]
Right, my comment about classproperties wasn't meant as a general indictment!

A couple of observations about native types:

* Instance-level data will be needed for `ModelType.native_type` to be equivalent to the specified model class. Otherwise it must be statically set to just `Model`.

* If `somefield.native_type = T`, it is only guaranteed that `isinstance(modelobj.somefield, T)`, whereas type *equivalence* is not guaranteed. I've thought of it sometimes: should `StringType` convert a custom subclass of `str` to `str`? Currently it doesn't do that, and even if we were to start normalizing types this way, it's pretty clear that at least `ModelType` should still leave subclass instances intact.

> Another question is whether we should enforce that this attribute exists and is not `None` in the `TypeMeta` class, and assert that the result of `to_primitive` matches the declared type. Or should we simply document this as a best practice?

I'd like to keep the attribute optional, at least initially. I'm trying my best to limit the amount of code existing applications need to change when migrating to v2.0.

Type enforcement would be nice but it seems wasteful to do it in production. We'll see how it performs. (This isn't the first time I've hoped Schematics had a global configuration module.)

**Comment by chadrik** [comment]
All of that sounds good to me.  I feel fairly comfortable doing the primitive type change, but I think there may be some nuances I may miss with the native types.  Let me know if you want me to take a stab at both or if you have time to do one or the other.

**Comment by lkraider** [comment]
Merged in #431
